File: src/mvgkit/common/camera.cc
#include "camera.h"

namespace mvgkit {
namespace common {

ArrayXf
RadialDistortionModel::getCoordCoeffs(const Array2Xf& imagePoints) const
{
  ArrayXf r2 = (imagePoints.matrix().colwise().squaredNorm()).array();
  ArrayXf r4 = r2.square();
  ArrayXf r6 = r4 * r2;
  ArrayXf coeffs = _coeffs[0] * r2 + _coeffs[1] * r4 + _coeffs[2] * r6 + 1.0;
  return coeffs;
}

ArrayXf
TangentialDistortionModel::getCoordCoeffs(const Array2Xf& imagePoints) const
{
  ArrayXf r2 = (imagePoints.matrix().colwise()).squaredNorm().array();
  ArrayXf x = imagePoints.row(0);
  ArrayXf y = imagePoints.row(1);
  ArrayXf xy = x * y;
  Array2Xf coeffs(2, imagePoints.cols());
  coeffs.row(0) = xy * _coeffs[0] * 2.0 + (r2 + x.square() * 2.0) * _coeffs[1];
  coeffs.row(1) = xy * _coeffs[1] * 2.0 + (r2 + y.square() * 2.0) * _coeffs[0];
  return coeffs;
}

CameraMatrix::CameraMatrix(float fx, float fy, float cx, float cy, float s)
{
  _coeffs << fx, fy, cx, cy, s;
}

CameraMatrix::CameraMatrix(const Matrix3f& mat)
{
  _coeffs << mat(0, 0), mat(1, 1), mat(0, 2), mat(1, 2), mat(0, 1);
}

Matrix3f
CameraMatrix::asMatrix() const
{
  Matrix3f K = Matrix3f::Identity();
  K(0, 0) = _coeffs[0];
  K(1, 1) = _coeffs[1];
  K(0, 2) = _coeffs[2];
  K(1, 2) = _coeffs[3];
  K(0, 1) = _coeffs[4];
  return K;
}

CameraMatrix
CameraMatrix::fromMatrix(const Matrix3f& mat)
{
  return CameraMatrix(mat(0, 0), mat(1, 1), mat(0, 2), mat(1, 2), mat(0, 1));
}

Array2Xf
CameraMatrix::project(const Array3Xf& points3d_C) const
{
  Array3Xf projected = asMatrix() * points3d_C.matrix();
  projected.topRows<2>().rowwise() /= projected.row(2);
  return projected.topRows<2>().array();
}

Array2Xf
CameraMatrix::projectToNormalizedImagePlane(const Array3Xf& points3d_C) const
{
  return points3d_C.topRows<2>().rowwise() / points3d_C.row(2);
}

Array2Xf
CameraMatrix::projectToPixels(const Array2Xf& normalizedImagePoints) const
{
  const auto& K = asMatrix();
  return ((K.block<2, 2>(0, 0) * normalizedImagePoints.matrix()).colwise() + K.block<2, 1>(0, 2))
    .array();
}

Array2Xf
CameraMatrix::unprojectToNormalizedImagePlane(const Array2Xf& sensorImagePoints) const
{
  const auto& K = asMatrix();
  return ((sensorImagePoints.matrix().colwise() - K.block<2, 1>(0, 2)) *
          K.block<2, 2>(0, 0).inverse())
    .array();
}

Vector5f
Camera::getDistortionCoefficients() const
{
  Vector5f coeffs;
  const Vector3f& rdcoeffs = _rdmodel.asArray();
  const Vector2f& tdcoeffs = _tdmodel.asArray();
  coeffs << rdcoeffs[0], rdcoeffs[1], tdcoeffs[0], rdcoeffs[2], tdcoeffs[1];
  return coeffs;
}

Camera::Camera(const CameraMatrix& cameraMatrix, const SE3f& pose_CW)
  : _K(cameraMatrix)
  , _pose_CW(pose_CW)
{
}

Camera::Camera(const CameraMatrix& cameraMatrix,
               const RadialDistortionModel& radial_distortion_model,
               const SE3f& pose_CW)
  : _K(cameraMatrix)
  , _rdmodel(radial_distortion_model)
  , _pose_CW(pose_CW)
{
}

Array2Xf
Camera::projectPoints(const Array3Xf& points3d_W, bool distort) const
{
  if (distort) {
    // TODO
  }
  Array3Xf points3d_C =
    (_pose_CW.rotationMatrix() * points3d_W.matrix()).colwise() + _pose_CW.translation();
  return _K.project(points3d_C);
}

} // common
} // mvgkit

File: src/mvgkit/common/camera.py
"""This module includes camera models such as a typical pinhole camera model.

TODO: Check for tilt distortion.
TODO: Implement orthographic projection
TODO: Implement image (un)-projection
"""

from enum import IntEnum


class ProjectionType(IntEnum):
    kPerspective = 0
    kOrthographic = 1


from pymvgkit_common import (  # noqa
    Camera,
    CameraMatrix,
    RadialDistortionModel,
    TangentialDistortionModel,
)

# @dataclass
# class PyCamera:
#     """
#     This class describes a pinhole camera model.
#     """

#     K: CameraMatrix
#     k: Optional[RadialDistortionModel] = RadialDistortionModel()
#     p: Optional[TangentialDistortionModel] = TangentialDistortionModel()
#     image_size: Optional[Tuple[int]] = (0, 0)

#     # Extrinsics, transforms the points in reference frame to camera frame.
#     T: Optional[SE3] = SE3()

#     def __post_init__(self):
#         self._P = self.K.as_matrix() @ self.T.as_augmented_matrix()

#     @property
#     def P(self):
#         """Camera projection matrix."""
#         return self._P

#     @staticmethod
#     def from_projection_matrix(self):
#         pass

#     def project_points(self, points_W, distort=False):
#         """Project points in world frame to image points."""
#         points_C = (self.T @ homogenize(points_W))[:, :3]
#         image_points = self.K.project(points_C)
#         if distort:
#             return self.distort_points(image_points)
#         return image_points

#     def distort_points(self, undistorted_image_points: np.ndarray):
#         """Distort the undistorted image points."""
#         normalized_image_points = self.K.unproject_to_normalized_image_plane(
#             undistorted_image_points
#         )
#         k_coeffs = self.k.get_coord_coeffs(normalized_image_points)
#         p_coeffs = self.p.get_coord_coeffs(normalized_image_points)
#         distorted = normalized_image_points * k_coeffs + p_coeffs
#         image_points = self.K.project_to_sensor_image_plane(normalized_image_points=distorted)
#         return image_points

#     def undistort_points(self, distorted_image_points: np.ndarray):
#         """Undistort the distorted image points."""
#         normalized_image_points = self.K.unproject_to_normalized_image_plane(distorted_image_points)
#         k_coeffs = self.k.get_coord_coeffs(normalized_image_points)
#         p_coeffs = self.p.get_coord_coeffs(normalized_image_points)
#         undistorted = (normalized_image_points - p_coeffs) / k_coeffs
#         image_points = self.K.project_to_sensor_image_plane(normalized_image_points=undistorted)
#         return image_points

#     def get_distortion_coefficients(self):
#         k_coeffs = self.k.as_array()
#         p_coeffs = self.p.as_array()
#         return np.asarray([k_coeffs[0], k_coeffs[1], p_coeffs[0], p_coeffs[1], k_coeffs[2]])

#     def get_optimal_camera_matrix(self, src_image_size, dest_image_size, alpha=1.0):
#         """Get the optimal matrix that can remove the unwanted black pixels after applying undistortion."""
#         raw_camera_matrix, roi = cv2.getOptimalNewCameraMatrix(
#             cameraMatrix=self.K.as_matrix(),
#             distCoeffs=self.get_distortion_coefficients(),
#             imageSize=src_image_size,
#             alpha=alpha,
#             newImgSize=dest_image_size,
#         )

#         return CameraMatrix.from_matrix(raw_camera_matrix), roi

#     def undistort_image(self, image: image_processing.Image, alpha: float = 1.0, crop=False):
#         new_K, roi = self.get_optimal_camera_matrix(image.size, image.size, alpha)
#         undistorted = cv2.undistort(
#             image.data,
#             self.K.as_matrix(),
#             self.get_distortion_coefficients(),
#             None,
#             new_K.as_matrix(),
#         )
#         if crop:
#             x, y, w, h = roi
#             undistorted = undistorted[y : y + h, x : x + w]
#         return image_processing.Image(undistorted, image.timestamp)


# @dataclass
# class OrthographicCamera:
#     T: SE3
